components/raftstore/src/store/worker/refresh_config.rs

```diff
@@ -0,0 +1,214 @@
+    pub router: BatchRouter<N, C>,
+    pub state: PoolState<N, C, H>,
+}
+
+impl<N, C, H> PoolController<N, C, H>
+where
+    N: Fsm,
+    C: Fsm,
+    H: HandlerBuilder<N, C>,
+{
+    #[allow(dead_code)]
+    pub fn new(router: BatchRouter<N, C>, state: PoolState<N, C, H>) -> Self {
+        PoolController { router, state }
+    }
+}
+
+impl<N, C, H> PoolController<N, C, H>
+where
+    N: Fsm + std::marker::Send + 'static,
+    C: Fsm + std::marker::Send + 'static,
+    H: HandlerBuilder<N, C>,
+{
+    pub fn decrease_by(&mut self, size: usize) {
+        for _ in 0..size {
+            if let Err(e) = self.state.fsm_sender.send(FsmTypes::Empty) {
+                error!(
+                    "failed to decrese thread pool";
+                    "decrease to" => size,
+                    "err" => %e,
+                );
+                return;
+            }
+        }
+    }
+
+    pub fn increase_by(&mut self, size: usize) {
+        let name_prefix = self.state.name_prefix.clone();
+        let mut workers = self.state.workers.lock().unwrap();
+        self.state.id_base += size;
+        for i in 0..size {
+            let handler = self.state.handler_builder.build(Priority::Normal);
+            let mut poller = Poller {
+                router: self.router.clone(),
+                fsm_receiver: self.state.fsm_receiver.clone(),
+                handler,
+                max_batch_size: self.state.max_batch_size,
+                reschedule_duration: self.state.reschedule_duration,
+                before_pause_wait: self.state.before_pause_wait,
+                joinable_workers: Some(Arc::clone(&self.state.joinable_workers)),
+            };
+            let props = tikv_util::thread_group::current_properties();
+            let t = thread::Builder::new()
+                .name(thd_name!(format!(
+                    "{}-{}",
+                    name_prefix,
+                    i + self.state.id_base,
+                )))
+                .spawn(move || {
+                    tikv_util::thread_group::set_properties(props);
+                    set_io_type(IOType::ForegroundWrite);
+                    poller.poll();
+                })
+                .unwrap();
+            workers.push(t);
+        }
+    }
+
+    fn cleanup_poller_threads(&mut self) {
+        let mut joinable_workers = self.state.joinable_workers.lock().unwrap();
+        let mut workers = self.state.workers.lock().unwrap();
+        let mut last_error = None;
+        for tid in joinable_workers.drain(..) {
+            if let Some(i) = workers.iter().position(|v| v.thread().id() == tid) {
+                let h = workers.swap_remove(i);
+                debug!("cleanup poller waiting for {}", h.thread().name().unwrap());
+                if let Err(e) = h.join() {
+                    error!("failed to join worker thread: {:?}", e);
+                    last_error = Some(e);
+                }
+            } else {
+                unreachable!()
```

Comment:
When shutting down and there are ongoing adjust pool size tasks, would it be possible that `BatchSystem::shutdown` run before `cleanup_poller_threads` and cause panic here?

Reply:
OK, let's omit the else branch to avoid shutdown and perform cleanup work first